File: analyze_xandel.py
import json
import csv
import os
from pathlib import Path

# Ler o JSON
json_path = r'c:\java\workspace\Gerador\metamodel\data\xandel\xandel-entities.json'
csv_dir = r'c:\java\workspace\Gerador\metamodel\data\xandel\carga-inicial'

with open(json_path, 'r', encoding='utf-8') as f:
    data = json.load(f)

entities = data.get('entities_continuation', [])

# Mapear CSVs disponíveis
csv_files = {}
for csv_file in Path(csv_dir).glob('*.csv'):
    csv_files[csv_file.stem.lower()] = csv_file

print("=" * 100)
print("RELATÓRIO DE COMPATIBILIDADE - JSON vs CSV")
print("=" * 100)
print()

# Estatísticas gerais
total_entities = len(entities)
total_csvs = len(csv_files)
print(f"Total de entidades no JSON: {total_entities}")
print(f"Total de CSVs disponíveis: {total_csvs}")
print()

# Análise por entidade
entities_with_csv = []
entities_without_csv = []
field_mismatches = []

for entity in entities:
    table_name = entity.get('tableName', '')
    entity_id = entity.get('id', '')
    entity_name = entity.get('name', '')

    print("-" * 100)
    print(f"ENTIDADE: {entity_name} (ID: {entity_id}, Tabela: {table_name})")
    print("-" * 100)

    # Buscar CSV correspondente
    csv_key = table_name.lower()

    if csv_key not in csv_files:
        print(f"❌ CSV NÃO ENCONTRADO para tabela '{table_name}'")
        entities_without_csv.append({
            'entity': entity_name,
            'table': table_name,
            'id': entity_id
        })
        print()
        continue

    entities_with_csv.append(entity_name)
    csv_path = csv_files[csv_key]

    # Ler cabeçalho do CSV
    try:
        with open(csv_path, 'r', encoding='utf-8') as csvfile:
            reader = csv.reader(csvfile)
            csv_columns = next(reader)
            csv_columns_lower = [col.lower() for col in csv_columns]
    except Exception as e:
        print(f"⚠️  ERRO ao ler CSV: {e}")
        print()
        continue

    print(f"✅ CSV ENCONTRADO: {csv_path.name}")
    print(f"   Colunas no CSV ({len(csv_columns)}): {', '.join(csv_columns)}")
    print()

    # Extrair campos da entidade
    fields = entity.get('fields', [])
    json_columns = []
    json_column_map = {}

    for field in fields:
        col_name = field.get('columnName', '')
        field_name = field.get('name', '')
        data_type = field.get('dataType', '')
        is_pk = field.get('primaryKey', False)
        is_required = field.get('required', False)
        reference = field.get('reference', {})

        json_columns.append(col_name)
        json_column_map[col_name.lower()] = {
            'name': field_name,
            'dataType': data_type,
            'isPK': is_pk,
            'required': is_required,
            'reference': reference
        }

    print(f"   Campos no JSON ({len(json_columns)}): {', '.join(json_columns)}")
    print()

    # Comparar campos
    json_set = set(col.lower() for col in json_columns)
    csv_set = set(csv_columns_lower)

    # Campos no JSON mas não no CSV
    missing_in_csv = json_set - csv_set
    # Campos no CSV mas não no JSON
    extra_in_csv = csv_set - json_set
    # Campos comuns
    common = json_set & csv_set

    if missing_in_csv or extra_in_csv:
        field_mismatches.append({
            'entity': entity_name,
            'table': table_name,
            'missing_in_csv': missing_in_csv,
            'extra_in_csv': extra_in_csv
        })

        if missing_in_csv:
            print(f"   ⚠️  CAMPOS NO JSON MAS NÃO NO CSV ({len(missing_in_csv)}):")
            for col in sorted(missing_in_csv):
                field_info = json_column_map.get(col, {})
                pk_marker = " [PK]" if field_info.get('isPK') else ""
                req_marker = " [OBRIGATÓRIO]" if field_info.get('required') else ""
                ref_marker = ""
                if field_info.get('reference'):
                    ref_entity = field_info['reference'].get('entity', '')
                    ref_marker = f" [FK -> {ref_entity}]"
                print(f"      - {col}{pk_marker}{req_marker}{ref_marker}")
            print()

        if extra_in_csv:
            print(f"   ⚠️  CAMPOS NO CSV MAS NÃO NO JSON ({len(extra_in_csv)}):")
            for col in sorted(extra_in_csv):
                original_col = next((c for c in csv_columns if c.lower() == col), col)
                print(f"      - {original_col}")
            print()
    else:
        print(f"   ✅ TODOS OS CAMPOS CORRESPONDEM ({len(common)} campos)")
        print()

    # Verificar relacionamentos
    relationships = entity.get('childEntities', [])
    if relationships:
        print(f"   📋 RELACIONAMENTOS (hasMany):")
        for rel in relationships:
            rel_entity = rel.get('entity', '')
            fk = rel.get('foreignKey', '')
            print(f"      - {entity_name} -> {rel_entity} (FK: {fk})")
        print()

    # Verificar foreign keys
    fk_fields = [f for f in fields if f.get('reference')]
    if fk_fields:
        print(f"   🔗 FOREIGN KEYS ({len(fk_fields)}):")
        for fk_field in fk_fields:
            col_name = fk_field.get('columnName', '')
            ref = fk_field.get('reference', {})
            ref_entity = ref.get('entity', '')
            ref_field = ref.get('field', '')
            on_delete = ref.get('onDelete', '')
            on_delete_str = f" [ON DELETE {on_delete}]" if on_delete else ""

            # Verificar se a FK está no CSV
            if col_name.lower() in csv_columns_lower:
                status = "✅"
            else:
                status = "❌"

            print(f"      {status} {col_name} -> {ref_entity}.{ref_field}{on_delete_str}")
        print()

print()
print("=" * 100)
print("RESUMO GERAL")
print("=" * 100)
print()

print(f"📊 Entidades com CSV correspondente: {len(entities_with_csv)}/{total_entities}")
print(f"❌ Entidades SEM CSV: {len(entities_without_csv)}/{total_entities}")
print(f"⚠️  Entidades com incompatibilidades de campos: {len(field_mismatches)}/{total_entities}")
print()

if entities_without_csv:
    print("ENTIDADES SEM CSV:")
    for ent in entities_without_csv:
        print(f"  - {ent['entity']} (tabela: {ent['table']})")
    print()

if field_mismatches:
    print("RESUMO DE INCOMPATIBILIDADES POR ENTIDADE:")
    for mismatch in field_mismatches:
        total_issues = len(mismatch['missing_in_csv']) + len(mismatch['extra_in_csv'])
        print(f"  - {mismatch['entity']} ({total_issues} discrepâncias)")
        if mismatch['missing_in_csv']:
            print(f"      • {len(mismatch['missing_in_csv'])} campos no JSON mas não no CSV")
        if mismatch['extra_in_csv']:
            print(f"      • {len(mismatch['extra_in_csv'])} campos no CSV mas não no JSON")
    print()

# CSVs sem entidade correspondente
csv_names = set(csv_files.keys())
entity_tables = set(e.get('tableName', '').lower() for e in entities)
orphan_csvs = csv_names - entity_tables

if orphan_csvs:
    print(f"📁 CSVs SEM entidade correspondente no JSON ({len(orphan_csvs)}):")
    for csv_name in sorted(orphan_csvs):
        print(f"  - {csv_name}.csv")
    print()

print("=" * 100)

File: analyze_xandel.js
const fs = require('fs');
const path = require('path');

// Paths
const jsonPath = path.join(__dirname, 'metamodel', 'data', 'xandel', 'xandel-entities.json');
const csvDir = path.join(__dirname, 'metamodel', 'data', 'xandel', 'carga-inicial');

// Read JSON
const jsonData = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
const entities = jsonData.entities_continuation || [];

// Read CSV files
const csvFiles = {};
fs.readdirSync(csvDir)
  .filter(file => file.endsWith('.csv'))
  .forEach(file => {
    const tableName = path.basename(file, '.csv').toLowerCase();
    csvFiles[tableName] = path.join(csvDir, file);
  });

console.log('='.repeat(100));
console.log('RELATÓRIO DE COMPATIBILIDADE - JSON vs CSV');
console.log('='.repeat(100));
console.log();

// Statistics
console.log(`Total de entidades no JSON: ${entities.length}`);
console.log(`Total de CSVs disponíveis: ${Object.keys(csvFiles).length}`);
console.log();

// Analysis
const results = {
  withCSV: [],
  withoutCSV: [],
  mismatches: [],
  csvWithoutEntity: []
};

// Helper to read CSV header
function readCSVHeader(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const firstLine = content.split('\n')[0].trim();
  return firstLine.split(';').map(col => col.trim());
}

// Analyze each entity
entities.forEach(entity => {
  const tableName = entity.tableName || '';
  const entityId = entity.id || '';
  const entityName = entity.name || '';
  const tableNameLower = tableName.toLowerCase();

  console.log('-'.repeat(100));
  console.log(`ENTIDADE: ${entityName} (ID: ${entityId}, Tabela: ${tableName})`);
  console.log('-'.repeat(100));

  // Check if CSV exists
  if (!csvFiles[tableNameLower]) {
    console.log(`❌ CSV NÃO ENCONTRADO para tabela '${tableName}'`);
    results.withoutCSV.push({ entity: entityName, table: tableName, id: entityId });
    console.log();
    return;
  }

  results.withCSV.push(entityName);
  const csvPath = csvFiles[tableNameLower];

  // Read CSV columns
  let csvColumns = [];
  try {
    csvColumns = readCSVHeader(csvPath);
    const csvColumnsLower = csvColumns.map(c => c.toLowerCase());

    console.log(`✅ CSV ENCONTRADO: ${path.basename(csvPath)}`);
    console.log(`   Colunas no CSV (${csvColumns.length}): ${csvColumns.join(', ')}`);
    console.log();

    // Extract fields from entity
    const fields = entity.fields || [];
    const jsonColumns = [];
    const jsonColumnMap = {};

    fields.forEach(field => {
      const colName = field.columnName || '';
      const fieldName = field.name || '';
      const dataType = field.dataType || '';
      const isPK = field.primaryKey || false;
      const isRequired = field.required || false;
      const reference = field.reference || null;

      jsonColumns.push(colName);
      jsonColumnMap[colName.toLowerCase()] = {
        name: fieldName,
        dataType,
        isPK,
        required: isRequired,
        reference
      };
    });

    console.log(`   Campos no JSON (${jsonColumns.length}): ${jsonColumns.join(', ')}`);
    console.log();

    // Compare columns
    const jsonSet = new Set(jsonColumns.map(c => c.toLowerCase()));
    const csvSet = new Set(csvColumnsLower);

    const missingInCSV = [...jsonSet].filter(c => !csvSet.has(c));
    const extraInCSV = [...csvSet].filter(c => !jsonSet.has(c));
    const common = [...jsonSet].filter(c => csvSet.has(c));

    if (missingInCSV.length > 0 || extraInCSV.length > 0) {
      results.mismatches.push({
        entity: entityName,
        table: tableName,
        missingInCSV,
        extraInCSV
      });

      if (missingInCSV.length > 0) {
        console.log(`   ⚠️  CAMPOS NO JSON MAS NÃO NO CSV (${missingInCSV.length}):`);
        missingInCSV.sort().forEach(col => {
          const fieldInfo = jsonColumnMap[col] || {};
          const pkMarker = fieldInfo.isPK ? ' [PK]' : '';
          const reqMarker = fieldInfo.required ? ' [OBRIGATÓRIO]' : '';
          let refMarker = '';
          if (fieldInfo.reference) {
            const refEntity = fieldInfo.reference.entity || '';
            refMarker = ` [FK -> ${refEntity}]`;
          }
          console.log(`      - ${col}${pkMarker}${reqMarker}${refMarker}`);
        });
        console.log();
      }

      if (extraInCSV.length > 0) {
        console.log(`   ⚠️  CAMPOS NO CSV MAS NÃO NO JSON (${extraInCSV.length}):`);
        extraInCSV.sort().forEach(col => {
          const originalCol = csvColumns.find(c => c.toLowerCase() === col) || col;
          console.log(`      - ${originalCol}`);
        });
        console.log();
      }
    } else {
      console.log(`   ✅ TODOS OS CAMPOS CORRESPONDEM (${common.length} campos)`);
      console.log();
    }

    // Check relationships
    const relationships = entity.childEntities || [];
    if (relationships.length > 0) {
      console.log(`   📋 RELACIONAMENTOS (hasMany):`);
      relationships.forEach(rel => {
        const relEntity = rel.entity || '';
        const fk = rel.foreignKey || '';
        console.log(`      - ${entityName} -> ${relEntity} (FK: ${fk})`);
      });
      console.log();
    }

    // Check foreign keys
    const fkFields = fields.filter(f => f.reference);
    if (fkFields.length > 0) {
      console.log(`   🔗 FOREIGN KEYS (${fkFields.length}):`);
      fkFields.forEach(fkField => {
        const colName = fkField.columnName || '';
        const ref = fkField.reference || {};
        const refEntity = ref.entity || '';
        const refField = ref.field || '';
        const onDelete = ref.onDelete || '';
        const onDeleteStr = onDelete ? ` [ON DELETE ${onDelete}]` : '';

        const status = csvColumnsLower.includes(colName.toLowerCase()) ? '✅' : '❌';
        console.log(`      ${status} ${colName} -> ${refEntity}.${refField}${onDeleteStr}`);
      });
      console.log();
    }

  } catch (error) {
    console.log(`⚠️  ERRO ao ler CSV: ${error.message}`);
    console.log();
  }
});

// Summary
console.log();
console.log('='.repeat(100));
console.log('RESUMO GERAL');
console.log('='.repeat(100));
console.log();

console.log(`📊 Entidades com CSV correspondente: ${results.withCSV.length}/${entities.length}`);
console.log(`❌ Entidades SEM CSV: ${results.withoutCSV.length}/${entities.length}`);
console.log(`⚠️  Entidades com incompatibilidades de campos: ${results.mismatches.length}/${entities.length}`);
console.log();

if (results.withoutCSV.length > 0) {
  console.log('ENTIDADES SEM CSV:');
  results.withoutCSV.forEach(ent => {
    console.log(`  - ${ent.entity} (tabela: ${ent.table})`);
  });
  console.log();
}

if (results.mismatches.length > 0) {
  console.log('RESUMO DE INCOMPATIBILIDADES POR ENTIDADE:');
  results.mismatches.forEach(mismatch => {
    const totalIssues = mismatch.missingInCSV.length + mismatch.extraInCSV.length;
    console.log(`  - ${mismatch.entity} (${totalIssues} discrepâncias)`);
    if (mismatch.missingInCSV.length > 0) {
      console.log(`      • ${mismatch.missingInCSV.length} campos no JSON mas não no CSV`);
    }
    if (mismatch.extraInCSV.length > 0) {
      console.log(`      • ${mismatch.extraInCSV.length} campos no CSV mas não no JSON`);
    }
  });
  console.log();
}

// Find CSVs without entities
const csvNames = new Set(Object.keys(csvFiles));
const entityTables = new Set(entities.map(e => (e.tableName || '').toLowerCase()));
const orphanCSVs = [...csvNames].filter(csv => !entityTables.has(csv));

if (orphanCSVs.length > 0) {
  console.log(`📁 CSVs SEM entidade correspondente no JSON (${orphanCSVs.length}):`);
  orphanCSVs.sort().forEach(csv => {
    console.log(`  - ${csv}.csv`);
  });
  console.log();
}

console.log('='.repeat(100));
